File: solutions/dart/atbash-cipher/1/lib/atbash_cipher.dart
// Solemnly dedicated to Shree DR.MDD
class AtbashCipher {
  String decode(String encrypted) =>
    encrypted.splitMapJoin('', onNonMatch: _flip);

  String encode(String original) =>
    decode(original.toLowerCase()).spacedBy(5);

  String _flip(String char) => switch (char) {
      'a' => 'z', 'b' => 'y', 'c' => 'x', 'd' => 'w', 'e' => 'v', 'f' => 'u', 'g' => 't',
      'h' => 's', 'i' => 'r', 'j' => 'q', 'k' => 'p', 'l' => 'o', 'm' => 'n',
      'n' => 'm', 'o' => 'l', 'p' => 'k', 'q' => 'j', 'r' => 'i', 's' => 'h',
      't' => 'g', 'u' => 'f', 'v' => 'e', 'w' => 'd', 'x' => 'c', 'y' => 'b', 'z' => 'a',
      '1' => '1', '2' => '2', '3' => '3', '4' => '4', '5' => '5',
      '6' => '6', '7' => '7', '8' => '8', '9' => '9', '0' => '0',
      _ => ''
    };
}

extension StringChunker on String {
  String spacedBy(int n) =>
    RegExp('.{1,$n}').allMatches(this).map((m) => m.group(0)).join(' ');
}

File: solutions/csharp/atbash-cipher/1/AtbashCipher.cs
// Dedicated to Shree DR.MDD
using System;
using System.Collections.Generic;
using System.Linq;

public static class AtbashCipher
{
    private const string AlphaSet = "abcdefghijklmnopqrstuvwxyz";
    private static readonly Dictionary<char, char> MirrorMap = AlphaSet.Zip(AlphaSet.Reverse()).ToDuoMap();

    public static string Encode(string input) =>
        input.ToLower()
             .Where(char.IsLetterOrDigit)
             .ToChop(5)
             .Select(set => set.Select(Transform).Unify())
             .Unify();

    public static string Decode(string encrypted) =>
        Encode(encrypted).Where(char.IsLetterOrDigit).Unify();

    private static char Transform(char symbol) => MirrorMap.GetValueOrDefault(symbol, symbol);

    private static IEnumerable<string> ToChop(this IEnumerable<char> stream, int len)
    {
        var data = stream.Unify();
        for (int step = 0; step < data.Length; step += len)
            yield return data.Substring(step, Math.Min(len, data.Length - step));
    }

    private static string Unify(this IEnumerable<char> sequence) => new(sequence.ToArray());

    private static string Unify(this IEnumerable<string> sections) => string.Join(" ", sections);

    private static Dictionary<TKey, TValue> ToDuoMap<TKey, TValue>(this IEnumerable<(TKey, TValue)> bundle) =>
        bundle.ToDictionary(el => el.Item1, el => el.Item2);
}
